File: src/main/java/Manager/Library.java
package Manager;

import java.io.File;
import java.io.IOException;
import java.util.ArrayList;

import Bean.AnimeClip;
import Bean.ResourceData;
import JavaFxPlus.Tool.ImageTool;
import JavaFxPlus.Util.FilterAbleList;
import Parser.SerializaOutput;
import Parser.SerializaParser;
import javafx.collections.FXCollections;
import javafx.collections.ObservableList;
import javafx.scene.image.Image;

public class Library {

	private static Library instance;

	public static Library getInstance()
	{
		if(instance==null)
		{
			instance=new Library();
		}
		return instance;
		
	}

	private FilterAbleList<ResourceData> list; 
	
	private int id=0;
	
	public Library() {
		ObservableList<ResourceData> list=FXCollections.observableArrayList();
		this.list=new FilterAbleList<>(list);
	}

	

	public void importImage(File file) {
		
		try {
			Image image=new Image(file.toURL().toString());
			String name=file.getName();
			name=name.substring(0, name.lastIndexOf("."));
			addToLibrary(name,ResourceType.TEXTURE,image);
			
		} catch (IOException e) {
			// TODO Auto-generated catch block
			e.printStackTrace();
		}
		
	}

	/**
	 * add a image to library
	 * @param name
	 * @param type
	 * @param data
	 */
	public void addToLibrary(String name,ResourceType type, Object data) {
		ResourceData resourceData=new ResourceData();
		resourceData.name=name;
		resourceData.type=type;
		resourceData.data=data;
		
		list.add(resourceData);
		
	}

	public ObservableList<ResourceData> getResources() {
		return list.getList();
	}

	/**
	 * remove all selected items
	 * @param selectedItems
	 */
	public void removeAll(ObservableList<ResourceData> selectedItems) {
		
		for (ResourceData resourceData : selectedItems) {
			remove(resourceData);
		}
		
	}

	private void remove(ResourceData resourceData) {
		
		list.remove(resourceData);
		
	}

	/**
	 * check whether a image with name has exist
	 * @param name
	 * @return
	 */
	public boolean hasSameName(String name) {
		
		return list.filter(d->d.name.equals(name)).size()>0;
	}

	/**
	 * slice a image and add to library
	 * @param image
	 * @param xAmount
	 * @param yAmount
	 * @param name
	 * @param jgzs 
	 * @param anime 
	 * @param zscdh  是否只生成动画不生成切片
	 */
	public void sliceAndAdd(Image image, int xAmount, int yAmount, String name, Boolean anime, int jgzs, Boolean zscdh) {
	
		
		int partWidth=(int) (image.getWidth()/xAmount);
		int partHeight=(int) (image.getHeight()/yAmount);
		
		
		
		for(int y=0;y<yAmount;y++)
		{
			//if need to generate a new anime
			AnimeClip animeClip = null;
			if(anime)
			{
				int maxFrame=xAmount*jgzs;
				animeClip=new AnimeClip(name+"anime"+y, maxFrame);
			}
			
			for(int x=0;x<xAmount;x++)
			{
				int xPos=x*partWidth;
				int yPos=y*partHeight;
				String autoNewName=name+x+y;
				
				
				Image sub=slice(image,xPos,yPos,partWidth,partHeight,autoNewName,zscdh);
				
				
				if(anime)
				{
					animeClip.insertFrame(x*jgzs+1, sub);
				}
				
			}
			
			if(anime)
			{
				addToLibrary(animeClip.getName(), ResourceType.MOVIECLIP, animeClip);
			}
		}
		
		
		
	}

	/**
	 * slice a image and add to library
	 * @param image
	 * @param xPos
	 * @param yPos
	 * @param partWidth
	 * @param partHeight
	 * @param autoNewName
	 * @param zscdh 
	 * @return 
	 */
	public Image slice(Image image, int xPos, int yPos, int partWidth, int partHeight, String autoNewName, boolean zscdh) {
		
		Image newImage=ImageTool.subImage(image,xPos,yPos,partWidth,partHeight);
		//if this do not exist,then add it
		if(hasSameName(autoNewName)==false&&!zscdh)
		{
			addToLibrary(autoNewName, ResourceType.TEXTURE, newImage);
		}
		
		
		return newImage;
		
	}



	public void fliterResource(String text) {
		list.filterList(d->d.name.startsWith(text));
	}



	public void createEmptyAnime() {
		
		String name="anime"+id++;
		addToLibrary(name,ResourceType.MOVIECLIP,new AnimeClip(name, 1));
			
		
	}



	public void outputToFile(File file) {
		
		//arraylist 才可以序列化 
		ArrayList<ResourceData> resourceDatas=new ArrayList<>(getResources());
		new SerializaOutput(resourceDatas,file);
		
		
		
	}



	public void inputFromFile(File file) {
		ArrayList<ResourceData> resourceDatas=new SerializaParser(file).toList();
		this.list.addAll(resourceDatas);
		
	}
}

File: src/main/java/Manager/Library.kt
package Manager

import java.io.File
import java.io.FileNotFoundException
import java.io.IOException
import java.util.ArrayList
import java.util.stream.Collectors

import com.google.gson.JsonIOException
import com.google.gson.JsonSyntaxException

import Bean.AnimeClip
import Bean.ResourceData
import JavaFxPlus.KotlinEx.toURLString
import JavaFxPlus.Tool.ImageTool
import JavaFxPlus.Util.FilterAbleList
import Parser.GsonOutput
import Parser.GsonParser
import Parser.SerializaOutput
import Parser.SerializaParser
import javafx.collections.FXCollections
import javafx.collections.ObservableList
import javafx.scene.image.Image

import java.util.stream.Collectors.*

object Library {

    private val list: FilterAbleList<ResourceData>

    private var id = 0

    init {
        val list = FXCollections.observableArrayList<ResourceData>()
        this.list = FilterAbleList(list)
    }


    fun importImage(file: File) {

        try {

            val image = Image(file.toURLString())
            var name = file.name
            name = name.substring(0, name.lastIndexOf("."))
            addToLibrary(name, ResourceType.TEXTURE, image)

        } catch (e: IOException) {
            // TODO Auto-generated catch block
            e.printStackTrace()
        }

    }

    /**
     * add a image to library
     * @param name
     * @param type
     * @param data
     */
    fun addToLibrary(name: String, type: ResourceType, data: Any) {
        val resourceData = ResourceData()
        resourceData.name = name
        resourceData.type = type
        resourceData.data = data

        list.add(resourceData)

    }

    val resources: ObservableList<ResourceData>
        get() = list.list

    /**
     * remove all selected items
     * @param selectedItems
     */
    fun removeAll(selectedItems: ObservableList<ResourceData>) =
            //shallow clone and remove
            selectedItems
                    .stream()
                    .collect(toList())
                    .stream()
                    .forEach { remove(it) }


    private fun remove(resourceData: ResourceData) = list.remove(resourceData)

    /**
     * check whether a image with name has exist
     * @param name
     * @return
     */
    fun hasSameName(name: String): Boolean {

        return list.filter { it.name == name }.size > 0
    }

    /**
     * slice a image and add to library
     * @param image
     * @param xAmount
     * @param yAmount
     * @param name
     * @param jgzs
     * @param anime
     * @param zscdh  是否只生成动画不生成切片
     */
    fun sliceAndAdd(image: Image, xAmount: Int, yAmount: Int, name: String, anime: Boolean, jgzs: Int, zscdh: Boolean) {


        val partWidth = (image.width / xAmount).toInt()
        val partHeight = (image.height / yAmount).toInt()



        for (y in 0..yAmount - 1) {
            //if need to generate a new anime
            var animeClip: AnimeClip? = null
            if (anime) {
                val maxFrame = xAmount * jgzs
                animeClip = AnimeClip(name + "anime" + y, maxFrame)
            }

            for (x in 0..xAmount - 1) {
                val xPos = x * partWidth
                val yPos = y * partHeight
                val autoNewName = name + x + y


                val sub = slice(image, xPos, yPos, partWidth, partHeight, autoNewName, zscdh)


                if (anime) {
                    animeClip!!.insertFrame(x * jgzs + 1, sub)
                }

            }

            if (anime) {
                addToLibrary(animeClip!!.name, ResourceType.MOVIECLIP, animeClip)
            }
        }


    }

    /**
     * slice a image and add to library
     * @param image
     * @param xPos
     * @param yPos
     * @param partWidth
     * @param partHeight
     * @param autoNewName
     * @param zscdh
     * @return
     */
    fun slice(image: Image, xPos: Int, yPos: Int, partWidth: Int, partHeight: Int, autoNewName: String, zscdh: Boolean): Image {

        val newImage = ImageTool.subImage(image, xPos, yPos, partWidth, partHeight)
        //if this do not exist,then add it
        if (hasSameName(autoNewName) == false && !zscdh) {
            addToLibrary(autoNewName, ResourceType.TEXTURE, newImage)
        }


        return newImage

    }


    fun fliterResource(text: String) {
        list.filterList { d -> d.name.startsWith(text) }
    }


    fun createEmptyAnime() {

        val name = "anime" + id++
        addToLibrary(name, ResourceType.MOVIECLIP, AnimeClip(name, 1))


    }


    fun outputToFileXLH(file: File) {

        //arraylist 才可以序列化
        val resourceDatas = ArrayList(resources)
        SerializaOutput(resourceDatas, file)


    }


    fun inputFromFileXLH(file: File) {
        val resourceDatas = SerializaParser(file).toList()
        this.list.addAll(resourceDatas)

    }


    fun inputFromFileGson(file: File) {

        var resourceDatas: ArrayList<ResourceData>? = null
        try {
            resourceDatas = GsonParser(file).toList()
        } catch (e: JsonIOException) {
            e.printStackTrace()
        } catch (e: JsonSyntaxException) {
            e.printStackTrace()
        } catch (e: FileNotFoundException) {
            e.printStackTrace()
        }

        this.list.addAll(resourceDatas)

    }


    fun outputToFileGson(file: File) {

        val resourceDatas = ArrayList(resources)
        try {
            GsonOutput(resourceDatas, file)
        } catch (e: IOException) {
            e.printStackTrace()
        }

    }


}
